lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -238,5 +241,178 @@
+                            //  3. Increase targetIndex by 1.
+                            ++targetIndex;
+                        }
+                        ++innerIndex;
+                    }
+                } else {
+                    //  vi. Else,
+                    //      1. If targetIndex >= 2^53-1, throw a TypeError exception.
+                    if (targetIndex >= 9007199254740991 /* 2^53 -1 */) {
+                        __chakraLibrary.raiseLengthIsTooBig("Array.prototype.flatMap");
+                    }
+                    //  2. Perform ? CreateDataPropertyOrThrow(target, ! ToString(targetIndex), element).
+                    __chakraLibrary.arrayCreateDataPropertyOrThrow(target, targetIndex, element);
+                    //  3. Increase targetIndex by 1.
+                    ++targetIndex;
+                }
+            }
+            //  d. Increase sourceIndex by 1.
+            ++sourceIndex;
+        }
+        //4. Return targetIndex.
+        return targetIndex;
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayFlat, function (depth) {
+        "use strict";
+        //1. Let O be ? ToObject(this value).
+        //2. Let sourceLen be ? ToLength(? Get(O, "length")).
+        let o, sourceLen;
+        
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            sourceLen = o.length;
+        } else {
+            if (this === null || this === undefined) {
+                __chakraLibrary.raiseThis_NullOrUndefined("Array.prototype.flat");
+            }
+            o = __chakraLibrary.Object(this);
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. Let depthNum be 1.
+        //4. If depth is not undefined, then
+        //5. Set depthNum to ? ToInteger(depth).
+        const depthNum = depth !== undefined ? __chakraLibrary.toInteger(depth) : 1;
+        //6. Let A be ? ArraySpeciesCreate(O, 0).
+        const A = __chakraLibrary.arraySpeciesCreate(o, 0);
+        //7. Perform ? FlattenIntoArray(A, O, sourceLen, 0, depthNum).
+        __chakraLibrary.FlattenIntoArray(A, o, sourceLen, 0, depthNum);
+        //8. Return A.
+        return A;
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayFlatMap, function (mapperFunction, thisArg) {
+        "use strict";
+        //1. Let O be ? ToObject(this value).
+        //2. Let sourceLen be ? ToLength(? Get(O, "length")).
+        let o, sourceLen;
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            sourceLen = o.length;
+        } else {
+            if (this === null || this === undefined) {
+                __chakraLibrary.raiseThis_NullOrUndefined("Array.prototype.flatMap");
+            }
+            o = __chakraLibrary.Object(this);
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. If IsCallable(mapperFunction) is false throw a TypeError exception
+        if (typeof mapperFunction != "function") {
```

Comment:
Nit: `!==` is generally preferable

Reply:
I copied that from Array.prototype.filter above - so if changing this one will change filter as well for consistency.

Reply:
Seems reasonable to me, thanks!

---
In reply to: [208360161](https://github.com/Microsoft/ChakraCore/pull/5573#discussion_r208360161) [](ancestors = 208360161)